cardano-db-sync/src/Cardano/DbSync/Util/Constraint.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.DbSync.Util.Constraint (
+  constraintNameEpochStake,
+  constraintNameReward,
+  dbConstraintNamesExists,
+  alterConstraints,
+  addRewardTableConstraint,
+  addEpochStakeTableConstraint,
+) where
+
+import Cardano.Db (ManualDbConstraints (..))
+import qualified Cardano.Db as DB
+import Cardano.Prelude (MonadIO, Proxy (..), ReaderT (runReaderT))
+import Control.Monad (when)
+import Control.Monad.Trans.Control (MonadBaseControl)
+import Database.Persist.Names (ConstraintNameDB (..), FieldNameDB (..))
+import Database.Persist.Postgresql (PersistEntity (..), SqlBackend)
+
+constraintNameEpochStake :: ConstraintNameDB
+constraintNameEpochStake = ConstraintNameDB "unique_epoch_stake"
+
+constraintNameReward :: ConstraintNameDB
+constraintNameReward = ConstraintNameDB "unique_reward"
+
+-- We manually create unique constraints to improve insert speeds when syncing
+-- This function checks if those constraints have already been created
+dbConstraintNamesExists :: MonadIO m => SqlBackend -> m ManualDbConstraints
+dbConstraintNamesExists sqlBackend = do
+  runReaderT queryRewardAndEpochStakeConstraints sqlBackend
+
+queryRewardAndEpochStakeConstraints ::
+  MonadIO m =>
+  ReaderT SqlBackend m ManualDbConstraints
+queryRewardAndEpochStakeConstraints = do
+  resEpochStake <- DB.queryHasConstraint constraintNameEpochStake
+  resReward <- DB.queryHasConstraint constraintNameReward
+  pure $
+    ManualDbConstraints
+      { dbConstraintRewards = resReward
+      , dbConstraintEpochStake = resEpochStake
+      }
+
+alterConstraints ::
+  forall m.
+  (MonadBaseControl IO m, MonadIO m) =>
+  ManualDbConstraints ->
+  ReaderT SqlBackend m ()
+alterConstraints ManualDbConstraints {..} = do
+  when dbConstraintRewards addRewardTableConstraint
```

Comment:
I think `unless` should be used instead of `when`. Also this function is never used, but I think it can, see comment below. It couls also be renamed to something like `addConstraintsIfNotExist`